soc: esp32: XIP removed from Espressif targets

wmrsouza opened the pull request:
This PR removes XIP configuration option from Espressif targets.

ESP hardware is not supporting what ZephyrRTOS calls the Execute-In-Place.
As an result it is causing some issues related to included Zephyr linker
files. To address this problem, XIP configuration option is being removed
from Espressif targets.

wmrsouza committed:
soc: esp32: XIP removed from Espressif targets

The way ESP32 XIP works (with MMU and cache) does no fit the way Zephyr XIP
is implemented, causing issues related to included Zephyr linker files.
Flash code still resides in flash for execution, but MMU/Cache handles it
in such way that XIP might not (or should not) be used with current Zephyr
approach. To address this problem, XIP configuration option is being
removed from Espressif targets.

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

raffarost commented:
Can you include ESP32C2 in the mix?

uLipe commented on soc/espressif/esp32c3/default.ld:
```diff
@@ -793,14 +802,6 @@ SECTIONS
   .flash.rodata_end : ALIGN(0x10)
   {
     . = ALIGN(4);
-
```
Is this related to this change? It would not deserve a proprr commit that explains why it is being removed, since the relationship to XIP seems to be not that explicit?

sylvioalves replied:
It is indeed related to XIP. Should stay here.
Background: ' __tdata_start' definition depends whether `XIP` is or no set. Current snippet was a workaround to fix TLs stuff.

dcpleung commented on arch/Kconfig:
```diff
@@ -122,7 +122,7 @@ config RISCV
 	select BARRIER_OPERATIONS_BUILTIN
 	select ARCH_HAS_THREAD_PRIV_STACK_SPACE_GET if USERSPACE
-	imply XIP
+	imply XIP if !SOC_FAMILY_ESPRESSIF_ESP32
```
I don't think this should be here. A particular SoC and its configs are built on top of architecture code, so they should not "leak" backwards.

sylvioalves replied:
@dcpleung `imply` XIP will force XIP no matter what. Do you recommend a way to disable it?

dcpleung replied:
Have you tried disabling it in `_defconfig`? That should act as an overlay which should override the default in kconfig files.

sylvioalves replied:
@wmrsouza Would you check that?

wmrsouza replied:
I'll check it.

dcpleung replied:
I guess the proper way to address this is to push `imply XIP` config into SoC configs and out of the arch layer.

sylvioalves replied:
> I guess the proper way to address this is to push `imply XIP` config into SoC configs and out of the arch layer.

This the same approach used by xtensa arch.

sylvioalves replied:
@dcpleung do you thing this approach is valid? (PR was updated moving `imply XIP` to SoCs).

dcpleung replied:
I think this is better as the arch should not dictate this kind of SoC/board specific configs.

wmrsouza committed:
arch: riscv: imply XIP config pushed to SoC level

'imply XIP' pushed from arch/Kconfig/'config RISCV' to riscv SoCs Kconfig
files to allow riscv SoCs having XIP enabled (or not) at SoC level

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

dcpleung commented:
Do we need RISC-V maintainers to sign off on this? I think we should at least let them know what is coming.

sylvioalves commented:
> Do we need RISC-V maintainers to sign off on this? I think we should at least let them know what is coming.

I think we should wait for it.

marekmatej commented:
Hi @tagunil, Thx for your interest in the topic.

There are several reasons why Zephyr's view of XIP view does not fit in the ESP32 world. In ESP32 the code/data are not available in the address space of the flash memory. Instead, we must set up caches and use virtual address spaces to be able to run/access the Flash code/data.

You can refer to ESP32-xx Technical Reference Manual, chapter "External Memory Address Mapping". The Zephyr side reference would be `zephyr/include/linker/common-rom/*`.

tagunil commented:
> Hi @tagunil, Thx for your interest in the topic.
> 
> There are several reasons why Zephyr's view of XIP view does not fit in the ESP32 world. In ESP32 the code/data are not available in the address space of the flash memory. Instead, we must set up caches and use virtual address spaces to be able to run/access the Flash code/data.
> 
> You can refer to ESP32-xx Technical Reference Manual, chapter "External Memory Address Mapping". The Zephyr side reference would be `zephyr/include/linker/common-rom/*`.

@marekmatej, thank you for the explanation!
So am I right to understand that instead of linking the code to the flash addresses, you do it the other way around and use MMU and/or caches to map the flash to the addresses you've linked the code to?
If you're not using the common XIP code paths because of that, when do you copy the .data section to the RAM then?
